### PR TITLE
Fix: Relations Input search not found

### PR DESCRIPTION
**Pull request by PatrickJira**
Lately my team just hit ~10k Article entry. 

We has been reported that users found some problem regarding RelationInput. They found RelationInput sometimes return empty result (No Relation Available) even if they typing the first word of title correctly, but after typing longer (2-3 words), they will still found it.

From investigation, we found that the RelationInput try to call API search with `$containsi` (from `findAvailable`) then filter again with `startsWith` (from useFilter)

Relations `findAvailable` filter with `$containsi`
https://github.com/strapi/strapi/blob/main/packages/core/content-manager/server/controllers/relations.js#L141-L143

`ComboBox` filter with `useFilter().startsWith`
https://github.com/strapi/design-system/blob/main/packages/primitives/src/components/Combobox/Combobox.tsx#L969-L975

### What does it do?

Add optional query param `_m`, which bring possibility to change filter from `$containsi` to `$startsWithi`
_don't know if the name and value is okay or not, please suggest_

### Why is it needed?

To make big records relations search less hassle. Better UX.

### How to test it?

1. Create Strapi project with --quickstart
2. Create a model named `Article`. with a field `title` type `string`
3. Create a model named `EditorsPick`, with a relation (1-to-many) to Articles
4. Create Article entry with these titles
- `A Koala 01`
- `B Koala 02`
- `C Koala 03`
- `D Koala 04`
- `E Koala 05`
- `F Koala 06`
- `G Koala 07`
- `H Koala 08`
- `I Koala 09`
- `J Koala 10`
- `Koala Article 11`

5. Create an EditorsPick entry, try to search for `Hello World 11` with only `Hello`

Before this fix, FE will call api search for 10 Articles with title that contains `Hello`, sort it ascending. Which will result in A-J Hello. Then filter again with startsWith `Hello`. Result in search not found.

This fix will match both side to same the type of filter

### CODE REVIEW

**Comment by PatrickJira** [comment]
`ComboBox` filter with `useFilter().startsWith`

https://github.com/strapi/design-system/blob/6233e279f773ccedd38dd179d2d3d21951bc1538/packages/primitives/src/components/Combobox/Combobox.tsx#L974

File: packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js [hunk 1]
```diff
@@ -168,6 +168,7 @@ export const useRelation = (cacheKey, { relation, search }) => {
     setSearchParams({
       ...options,
       _q: term,
+      _m: 'startsWith',
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js:171]: would it make more sense to make `_m = startsWithi`?

File: packages/core/content-manager/server/controllers/relations.js [hunk 2]
```diff
@@ -139,7 +139,11 @@ module.exports = {
 
     // searching should be allowed only on mainField for permission reasons
     if (_q) {
-      addFiltersClause(queryParams, { [mainField]: { $containsi: _q } });
+      if(query._m && query._m == 'startsWith') {
+        addFiltersClause(queryParams, { [mainField]: { $startsWithi: _q } });
+      } else {
+        addFiltersClause(queryParams, { [mainField]: { $containsi: _q } });
+      }
```
Comment (joshuaellis) [inline packages/core/content-manager/server/controllers/relations.js:146]: If you think it makes sense to change `_m` then you can reduce this code to just be:

```suggestion
      if(_q) {
        addFiltersClause(queryParams, { [mainField]: { [`$${_m}`]: _q } });
      }
```

This means by default the controller can preserve the default behaviour, but the admin panel requests something different? 🤔

**Comment by PatrickJira** [comment]
Changes as @joshuaellis suggested. Please review

**Review by joshuaellis — APPROVED** [review]
niceeee

**Comment by jhoward1994** [comment]
Nice thanks for this! I'm just wondering a couple of things:

- Why are we using `_m` as the param name? To me that naming doesn't make much sense as we are changing a filter (_filter ?). Maybe there is something I'm missing?
- Should we somehow lock the value of `_m` down to only the filters accepted by the entity service here - https://docs.strapi.io/dev-docs/api/entity-service/filter ?

wdyt @joshuaellis @PatrickJira

**Comment by PatrickJira** [comment]
Hi @jhoward1994, 

I don't know if there are any preserved keywords for get params. So I'm using `_m` which stand for mode, and use underscore to less likely to hit some preserved keyword, also same pattern along side the existing one `_q`.

I'm more than happy to change if anyone can make sure we can use actual word like `searchType`, `searchOperand` or anything more readable without problems.

Yes, If you have suggestion, please do

On the second bullet. Yes, totally agree with that. Is there any existing helper/service for that? or do I only need add checking for `null | undefined`, `startsWithi` and `containsi` for now?

**Comment by PatrickJira** [comment]
@jhoward1994 @joshuaellis Any suggestion on `_filter` validation/limitation?

**Comment by PatrickJira** [comment]
Validation added, please review.

As I need to append `$` to the `_filter` to use with `isOperator`, it's looks ugly/inconsistency. So I change to send from Admin Panel instead. It's looks cleaner.

And as I use `isOperatorOfType` from @strapi/utils, which might be better to check only for `where` operator

**Comment by PatrickJira** [comment]
Test were failing, don't really know what to fix. Any suggestion?

**Comment by Boegie19** [comment]
Looking at the code changes I think you maybe had an unstable version of main. since your code changes should not change how react renders on the front end.

**Comment by PatrickJira** [comment]
What should I do? Rebase to which branch/tags? Just hit the update branch button days ago, but that not trigger the test to rerun tho.

**Comment by joshuaellis** [comment]
> Test were failing, don't really know what to fix. Any suggestion?

have you ran the tests locally? I imagine the `useRelation` hook checks what params it sends etc.

**Comment by PatrickJira** [comment]
Test Updated add `_filter: '$startsWithi'` to expected params

**Review by joshuaellis — APPROVED** [review]

**Pull request merged**